### PR TITLE
Add img2img and inpaint stable diffusion ORT pipelines

### PR DESCRIPTION
**Pull request by echarlaix**
Add `ORTStableDiffusionImg2ImgPipeline` and `ORTStableDiffusionInpaintPipeline` classes to load SD ONNX models and run inference with ONNX Runtime for image-to-image and  image inpainting tasks. In this PR we also enable the saving and loading of each model's subcomponent config

### CODE REVIEW

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -48,32 +50,48 @@ def _encode_prompt(
...
+            prompt_embeds (`np.ndarray`, *optional*):
+                Pre-generated text embeddings. Can be used to easily tweak text inputs, *e.g.* prompt weighting. If not
+                provided, text embeddings will be generated from `prompt` input argument.
+            negative_prompt_embeds (`np.ndarray`, *optional*):
+                Pre-generated negative text embeddings. Can be used to easily tweak text inputs, *e.g.* prompt
+                weighting. If not provided, negative_prompt_embeds will be generated from `negative_prompt` input
+                argument.
         """
-        batch_size = len(prompt) if isinstance(prompt, list) else 1
+        if prompt is not None and isinstance(prompt, str):
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion.py:61]:
```suggestion
        if isinstance(prompt, str):
```
Should be enough?

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -48,32 +50,48 @@ def _encode_prompt(
...
+            batch_size = 1
+        elif prompt is not None and isinstance(prompt, list):
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion.py:63]:
```suggestion
        elif isinstance(prompt, list):
```

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -124,28 +192,86 @@ def __call__(
...
+                [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
+            return_dict (`bool`, *optional*, defaults to `True`):
+                Whether or not to return a [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] instead of a
+                plain tuple.
+            callback (`Callable`, *optional*):
+                A function that will be called every `callback_steps` steps during inference. The function will be
+                called with the following arguments: `callback(step: int, timestep: int, latents: torch.FloatTensor)`.
+            callback_steps (`int`, *optional*, defaults to 1):
+                The frequency at which the `callback` function will be called. If not specified, the callback will be
+                called at every step.
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion.py:253]: nit: a bit of a pain but the docstring does not follow our format with the "*optional*".

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_img2img.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        negative_prompt_embeds: Optional[np.ndarray] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, np.ndarray], None]] = None,
+        callback_steps: int = 1,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
+
+        Args:
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion_img2img.py:131]: Same comment.

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_inpaint.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+        negative_prompt_embeds: Optional[np.ndarray] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, np.ndarray], None]] = None,
+        callback_steps: int = 1,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
+
+        Args:
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion_inpaint.py:131]: Same comment

File: setup.py [hunk 6]
```diff
@@ -57,6 +57,7 @@
     "exporters": ["onnx<1.14.0", "onnxruntime", "timm"],
     "exporters-gpu": ["onnx<1.14.0", "onnxruntime-gpu", "timm"],
     "exporters-tf": ["tensorflow>=2.4,<2.11", "tf2onnx", "onnx", "onnxruntime", "timm", "h5py", "numpy<1.24.0"],
+    "diffusers": ["diffusers"],
```
Comment (michaelbenayoun) [inline setup.py:60]: Do we test anywhere in the stable diffusion ONNX pipelines that diffusers is available (if needed)?

File: tests/onnxruntime/test_stable_diffusion_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+
+def _generate_random_inputs():
+    inputs = {
+        "prompt": "sailing ship in storm by Leonardo da Vinci",
+        "num_inference_steps": 3,
+        "guidance_scale": 7.5,
+        "output_type": "numpy",
+    }
+    return inputs
```
Comment (michaelbenayoun) [inline tests/onnxruntime/test_stable_diffusion_pipeline.py:55]: It's not random no?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 8]
```diff
@@ -106,24 +118,46 @@ def __init__(
...
         self.safety_checker = None
+
         sub_models = {
             DIFFUSION_MODEL_TEXT_ENCODER_SUBFOLDER: self.text_encoder,
             DIFFUSION_MODEL_UNET_SUBFOLDER: self.unet,
             DIFFUSION_MODEL_VAE_DECODER_SUBFOLDER: self.vae_decoder,
+            DIFFUSION_MODEL_VAE_ENCODER_SUBFOLDER: self.vae_encoder,
         }
+
+        # Modify config to keep the resulting model compatible with diffusers pipelines
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:141]: I don't understand what this means. Is it: if we use `save_pretrained` on this pipeline, the resulting model folder will be loadable in diffusers pipeline?

**Review by fxmarty — APPROVED** [review]
Not sure to understand all the changes, but LGTM

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py (hunk 1 above)
Comment (echarlaix) [reply optimum/pipelines/diffusers/pipeline_stable_diffusion.py:61]: Good catch !

File: optimum/onnxruntime/modeling_diffusion.py (hunk 8 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:141]: Yes it allows the resulting model to be compatible with the `diffusers` pipelines (such as `OnnxStableDiffusionPipeline`) which use the model config to determine how the model should be loaded (it needs to be loaded with diffusers `OnnxRuntimeModel`).

File: optimum/pipelines/diffusers/pipeline_stable_diffusion_img2img.py (hunk 4 above)
Comment (echarlaix) [reply optimum/pipelines/diffusers/pipeline_stable_diffusion_img2img.py:132]: Done !

File: setup.py (hunk 6 above)
Comment (echarlaix) [reply setup.py:60]: Yes we do : https://github.com/huggingface/optimum/blob/add-img2img/optimum/onnxruntime/__init__.py#L76 if `diffusers` is not installed when instantiating an `ORTStableDiffusionPipeline` an `ImportError` will be raised

File: optimum/onnxruntime/modeling_diffusion.py [hunk 9]
```diff
@@ -52,11 +60,12 @@
 logger = logging.getLogger(__name__)
 
 
-class ORTStableDiffusionPipeline(ORTModel, StableDiffusionPipelineMixin):
+class ORTStableDiffusionPipelineBase(ORTModel):
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:63]: each pipeline is a subclass of [`ConfigMixin`](https://github.com/huggingface/optimum/blob/add-img2img/optimum/pipelines/diffusers/pipeline_utils.py#L21) from `diffusers`, this is useful to correctly load each subcomponent of the pipeline like the tokenizer or  scheduler

File: optimum/onnxruntime/modeling_diffusion.py (hunk 8 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:141]: @patrickvonplaten, what do you think ?

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 10]
```diff
@@ -147,6 +153,128 @@ def __call__(
...
+        num_images_per_prompt: int = 1,
+        eta: float = 0.0,
+        generator: Optional[np.random.RandomState] = None,
+        latents: Optional[np.ndarray] = None,
+        prompt_embeds: Optional[np.ndarray] = None,
+        negative_prompt_embeds: Optional[np.ndarray] = None,
+        output_type: str = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, np.ndarray], None]] = None,
+        callback_steps: int = 1,
```
Comment (patrickvonplaten) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion.py:200]: (nit) These are very edge-casy use cases that are probably not working in ONNX and/or are not needed
```suggestion
```

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py [hunk 11]
```diff
@@ -210,25 +349,36 @@ def __call__(
...
-        # it seems likes there is a strange result for using half-precision vae decoder if batchsize>1
-        image = np.concatenate(
-            [self.vae_decoder(latent_sample=latents[i : i + 1])[0] for i in range(latents.shape[0])]
-        )
+        if output_type == "latent":
+            image = latents
+            has_nsfw_concept = None
+        else:
+            latents = 1 / self.vae_decoder.config.get("scaling_factor", 0.18215) * latents
+            # it seems likes there is a strange result for using half-precision vae decoder if batchsize>1
```
Comment (patrickvonplaten) [inline optimum/pipelines/diffusers/pipeline_stable_diffusion.py:357]: Interesting! Not sure if it's related to this: https://huggingface.slack.com/archives/C03UQJENJTV/p1688029497835759

File: optimum/onnxruntime/modeling_diffusion.py (hunk 8 above)
Comment (patrickvonplaten) [reply optimum/onnxruntime/modeling_diffusion.py:141]: I think long-term we want to deprecate `OnnxStableDiffusionPipeline` from `diffusers` actually, but for now I think this makes sense!

**Review by patrickvonplaten — APPROVED** [review]
Looks cool!

File: optimum/pipelines/diffusers/pipeline_stable_diffusion.py (hunk 10 above)
Comment (echarlaix) [reply optimum/pipelines/diffusers/pipeline_stable_diffusion.py:200]: Just added a fix for callbacks and added a test in [`7864a`](https://github.com/huggingface/optimum/pull/1121/commits/7864a43d65f8ea746493c1a5cffc00795048b5d3). Let me know if you think we should test additional test (used this [one](https://github.com/huggingface/diffusers/blob/v0.17.1/tests/pipelines/stable_diffusion/test_stable_diffusion.py#L878))

File: optimum/onnxruntime/modeling_diffusion.py (hunk 8 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:141]: Yes will remove it once the `OnnxStableDiffusionPipeline` is deprecated

**Pull request merged**